### ISSUE TITLE
Unhappy about `stack test --test-arguments`

### ISSUE THREAD

**Issue by benma** [commercialhaskell/stack#2210]
Hi

I wonder if there is a way to improve the long winded way to pass test arguments.

It took me a good 10 minutes including my best google-fu to find out how to run only a subset of my hspec tests (I am new to hspec). The end result is:

`stack test --test-arguments "-m <filter>"`

This is pretty long-winded, and when I return to using Haskell and Hspec after some weeks, I have to look it up again. Repeating this in custom bash functions or Makefiles is also not very nice. 

I think Haskell as a whole would benefit if common operations like this would be easier to type and remember. Take Go, which focuses on simple usage and batteries included:

`go test -run <filter>`

Maybe stack could have specific subcommands for testing frameworks like hspec, so one could run `stack hspec -m <filter>` or similar (and all other args that hspec is taking

I am aware that stack might not be the optimal place to cater to individual packages, but on the other hand, stack is more and more the goto tool for Haskell development, and I think it is worth including more batteries to make Haskell and everything around it even easier to use.

Thoughts?

**Comment by mgsloan** [comment]
I agree, I've often thought the same.

I can think of a couple modifications that would make it quite short:

* `stack test -- -m <filter>` - this would cause `stack test` to be more than just an alias for `stack build --test`.  It would also give a meaning to what comes after `--`, as used in `stack exec -- cmd`. 

* This might be a crazy idea, but what if `stack test "-m <filter>"` worked?

**Comment by benma** [comment]
What about a combination of your two proposals? Every flag that does not belong to `stack test` gets passed as test arguments, and also everything after `--`. If there is a clash of `stack test` flags and test arguments, you can move it after `--`, but for most flags, that shouldn't be a problem.

That might be a bit hacky as it is not forwards compatible, so I am would also be happy if only your first proposal would be implemented.

Apparently, @gregwebs also had the same idea:
https://github.com/commercialhaskell/stack/issues/172#issuecomment-111135123. There is some discussion on it.

**Comment by gregwebs** [comment]
@benma The idea of automatically passing flags to tests that do not belong to `stack test` creates fuzzy semantics. Now I have to remember what flags belong to `stack test` and understand collision behaviour where everything works but my flag doesn't have the effect I thought it would. When you upgrade stack and a new flag is added, your test script can break.

**Comment by gregwebs** [comment]
I am also unhappy with `--test-arguments` because now I have to deal with quoting things (that could need quotes!). It was decided that there was a principal at stake that must be interpreted a certain way and as a consequence it destroyed all ease of usability. Nothing bad will happen from using `--` and as per my last comment it makes CLI semantics very clear, and it is very easy to type.

**Comment by mgsloan** [comment]
Yup, I'm in favor of implementing the `stack bench -- args` and `stack test -- args`.

**Comment by benma** [comment]
Great. Sorry I am new here, is this now greenlighted, or are there other people who need to vote?

> Yup, I'm in favor of implementing the `stack bench -- args` and `stack test -- args` approach. 

Would this replace `--test-arguments` or would we keep that for backwards compatibility? I.e. provide the flags using either style, but not both?

**Comment by gregwebs** [comment]
Have to keep it for backwards compatibility. Either merging the 2 or throwing an error when they are both used seems fine to me.

**Comment by mgsloan** [comment]
Yup, we should keep both anyway so that you can do `stack build --test --bench --test-arguments a --bench-arguments b`.

> Sorry I am new here, is this now greenlighted, or are there other people who need to vote?

Consider it greenlighted!  We don't really have a set process for that, I'd say the amount of discussion and consensus required is related to the overall size and impact of the change.  Of course, people should feel free to disagree with a particular decision and discuss it.

For a backwards compatible CLI convenience with little downside, I feel quite comfortable saying we should do it.  The discussion in #172 isn't "let's not do the -- thing", instead it's "Lets do this reasonable thing and circle back to the -- later".  Now we are circling back :)

**Comment by benma** [comment]
> I can think of a couple modifications that would make it quite short:
>     stack test -- -m <filter> - this would cause stack test to be more than just an alias for stack build --test. It would also give a meaning to what comes after --, as used in stack exec -- cmd.

@mgsloan, @gregwebs  unfortunately, this doesn't work as smoothly as expected. The problem is that `stack test [TARGET]` takes a list of targets, and everything after ` -- ` is treated as positional arguments by optparse-applicative.

[From optparse-applicative](https://hackage.haskell.org/package/optparse-applicative-0.14.0.0):
> However, parsers always accept a special argument: --. When a -- is found on the command line, all the following words are considered by argument parsers, regardless of whether they start with - or not.

So `stack test target1 target2 -- --foo` would result in the targets `["target1", "target", "--foo"]`, leaving nothing for the test arguments.

The trick with the ` -- ` delimiter works in `stack exec` more or less by accident, because `stack exec` has only one positional argument, so it's unambiguous.

I am still thinking about what can be done about this, but so far, I couldn't come up with anything nice.

Thoughts?

**Comment by benma** [comment]
Huw Campbell, author of optparse-applicative, seems to have faced the same problem, and chose to pass along all unrecognized flags.

http://www.huwcampbell.com/posts/2017-06-09-announcing-optparse-0.14.html

> Fixing this issue inspired me to also alleviate a pain point which my colleagues and I had been having regarding mixing arguments and wrapping commands. At Ambiata, we created a build tool called mafia which acts as an opinionated wrapper around cabal, providing a nix like package cache, safe sandboxing, and submodule handling amongst other things.

> When using it though, to pass an option to cabal underneath, the user would need to first use `--` to tell optparse to treat all remaining options as arguments, and the argument list would be sent to cabal. Now, with the forwardOptions builder attached when creating the parserInfo structures, any unrecognised options to mafia can be passed through as is. I believe this will actually help in quite a few places where people are using mixed environments

@HuwCampbell I'd appreciate your thoughts, as you seem to have put in a lot of thought around this issue. 

Related: https://github.com/pcapriotti/optparse-applicative/issues/269

**Comment by benma** [comment]
Another thing we could do is to use a different separator than ` -- `, in which case it could easily be disambiguated. To be consistent, this would need to be used also in `stack exec` etc., so it might be not a welcome change to everyone who already trained their muscle memory.

**Comment by HuwCampbell** [comment]
That's possible, to get it to work well though (collecting flags as well after it was mentioned) you would have to use a subparser which only matches it in the front of an alternative which holds the arguments which come first (it would need an OnlyArguments policy as well. It would be an ugly hack.

**Comment by benma** [comment]
Ugly indeed!

**Comment by kirelagin** [comment]
Oh my god, `--test-arguments` is so terrible, I have been fighting with it for an hour or so already. I am trying to do the simplest thing possible – a shell script that does some setting up and then calls `stack test` passing all its arguments to the test runner. I’m starting to thing it’s almost impossible to make it work reliably with spaces.

After fighting with bash (😠) I settled on `"${*@Q}"` which is a string with each individual argument helpfully quoted by bash. I was 95% this would work, but, apparently instead of simply passing the single `--test-arguments` argument to the shell, stack tries to interpret it itself or god know what happens, but the problem is that bash’s `@Q` quotes arguments by single quotes and then #3091 happens.

It’s starting to look more and more like I’ll have to write an actual Python script that will be requoting the arguments. Wow.

**Comment by kirelagin** [comment]
In case it will help anyone, I ended up doing:

```
# Wrap every argument in single quotes, because that’s what stack wants :/
args=( "${@/#/\"}" ); args=( "${args[@]/%/\"}" )
stack test --test-arguments "${args[*]}"
```

I doubt it will work reliably, especially will all kinds of different quotation marks, and whitespace, and glob expansion, etc., but it seems to do what is expected in our use cases (i.e., mostly, `--match` with spaces).

**Comment by tuomohopia** [comment]
Sooo... what's the way to do this now? Say I want to match a test like this `it "POST /batteries/edit/:id edits the battery when authorized" $ ...`, how would I go on about this? 

I'm on WIN10.

**Comment by asarkar** [comment]
To add to @kirelagin comment about test filtering, using his shell script option, in order to run all tests from module `abc`, pass `-m abc`. If the module name contains whitespace, pass `-m "abc def"`. Note that these are substring matches, for exact match, use `-m /abc/` or `-m "/abc def/"`.

**Comment by s5k6** [comment]
Hi,

I'd like to share my thoughts on making `--test-arguments` correspond
to shell syntax a bit more.

I find the way `stack test` passes command line arguments to a test
suite unfortunate.  It implements its own notion of quoting, which of
curse will conflict with the conventions of the calling shell, what
ever that may be.

I would appreciate a means to pass arguments to a test suite in a way
that is compatible to how arguments are passed to a “usual” executable
**without** the introduction of additional quoting:

The list of arguments as understood by the shell should correspond to
the list of strings returned by `getArgs :: IO [String]`.

It should not play a role whether you're on Linux or not.  I assume
every shell has its natural way of passing a list of strings to an
executable, and that word-splitting should be used.

I see multiple ways out of this: The following examples all use bash
syntax (following the `$` prompt), and should all make `getArgs`
return the two-string list

    ["one", "two words"]

 1. Use `--` to separate test arguments from stack arguments, as has
    already been suggested.  This is equivalent to require that `--ta` is
    the last argument passed to `stack test`:

        $ stack test --file-watch --ta one 'two words'

    Using `--` to indicate “end of options” is also a known idiom in
    the Linux world.

    As has been pointed out above, this may conflict with other `stack
    test` options that take multiple arguments, e.g.,
    `--haddock-arguments` and `--benchmark-arguments`.  I do not know
    whether they may occur at the same time as `--test-arguments`.
    What about GHC compiler options?

 2. Devise a syntax to denote the beginning and the ending of a list
    of arguments:

        $ stack test --ta-MARK one 'two words' MARK --file-watch

    The idea is that the string `MARK`, *appended* to `--ta-` can be a
    somewhat arbitrary string, **chosen by the calling user**, whose
    next occurrence on the command line (as solitary argument) denotes
    the end of test arguments.  Thus, the following commands would all
    pass the same arguments to `getArgs` as the one above:

        $ stack test --ta-x one 'two words' x  --file-watch
        $ stack test --ta-: one 'two words' : --file-watch
        $ stack test --file-watch --ta one 'two words'

    Note, that the last line uses `--ta` instead of `--ta-` to
    indicate the “all the rest” semantics (like variant 1).  I'm not
    sure if it makes sense to enforce the presence of the ending
    marker, I just wanted to point out the possibility.

    A similar approach is taken for [GHCs Runtime system options][1],
    which are bracketed between `+RTS` and `-RTS`.

    A similar approach is taken by the [GNU `find` utility][2], where an
    executable command and its arguments are bracketed between `-exec`
    and `;` (the latter needs quoting on the shell).

 3. Individually mask every single argument that is to be passed to
    the test suite:

        $ stack test --file-watch --ta one --ta 'two words'

    I.e., every `--ta` consumes exactly one command line argument, and
    passes it on as one single entry in the list returned by
    `getArgs`, without further ado.  Verbose, but IMHO better than to
    have stack think about quotes passed inside an argument.

Variations and combinations of these are possible.  E.g., `--ta` could
just add its single following argument to the list (variant 3), and
`--ta-` could start a list (variant 2), thus

    $ stack test --ta a --file-watch --ta-: b 'c d' : --stack-argument --ta e f

could make `getArgs` return `["a", "b", "c d", "e"]`, using `:` as
list-delimiter, and passing `--stack-argument` and `f` to stack
directly.

I do prefer variant 2.  In general, this could be extended to all
list-valued command line arguments (`--haddock-arguments` and
`--benchmark-arguments`, etc.).  The approach nicely lends itself to
nesting, since the end-of-list marker can be chosen arbitrarily every
time it is used.



[1]: https://downloads.haskell.org/ghc/9.12.1/docs/users_guide/runtime_control.html
[2]: https://www.gnu.org/software/findutils/manual/html_mono/find.html#Run-Commands

**Comment by mpilgrem** [comment]
@s5k6, thanks for sharing. It is a complicated topic because there is:
* what the shell passes to `optparse-applicative` (the popular package that Stack uses to parse the command line) (neither `optparse-applicative` or Stack has any knowledge of, or control over, that);
* what `optparse-applicative` passes on to Stack (the Stack project has no control over that); and
* what Stack does with what it receives from `optparse-applicative`.

I think it is unlikely that Stack will move away from depending on the `optparse-applicative` package; that would involve a huge amount of re-plumbing and the package's conventions are well understood.

So, if you have a simple test executable:
~~~haskell
import System.Environment ( getArgs )

main :: IO ()
main = getArgs >>= print
~~~

and you want to pass `["one", "two words"]`, in PowerShell you have to do:
~~~text
stack test --ta 'one "two words"'
~~~

* the shell passes `--ta`, `one "two words"`;
* `optparse-applicative` interprets `one "two words"` as the value of the `--ta` option; and
* Stack interprets that as a desire for two arguments for the test executable: `one`, and `two words`. 

On PowerShell, the problem with:
~~~text
stack test --ta one "two words"
~~~

is:
* the shell passes `--ta`, `one`, `two words`;
* `optparse-applicative` interprets `one` as the value of the `--ta` option and `two words` as a separate argument; and
* Stack interprets `one` as the desired argument for the test exectuable and the further argument (`two words`) as the desired target.

If you try:
~~~text
stack test --ta one --ta "two words"
~~~

* the shell passes `--ta`, `one`, `--ta`, `two words`;
* `optparse-applicative` interprets one `--ta` value of `one` and a second `--ta` value of `two words`; and
* Stack will interpret the first `one` value as the desired first argument of the test exectuable and the second `two words` value as being a desire for two further arguments for the test executable: `two`, and `words`.

It is complicated, principally because you have to know what your shell is doing with the command line and what it is actually passing on to the executable. What `optparse-applicative`, and then Stack, are doing with what they receive is documented - but I can see more examples online would help users.

The `optparse-applicative` package does not provide anything similar to the GHC runtime system's `+RTS ... -RTS`.

The `optparse-applicative` package does support `-- <symbols>` at the end of a command line, so Stack could have a grammar to interpret:
~~~text
stack build ... -- <symbols>
~~~

but that grammar would need, I think, to cater for both `--test-arguments` and `benchmark-arguments`.

**Comment by s5k6** [comment]
Thanks for answering @mpilgrem.  I had already figured out how to pass `["one", "two words"]` with the current implementation, sorry if that was not clear.  I've just used that as an example to illustrate what I'd consider a more convenient way to do it.

Unfortunately, I do not know the precise mechanics by which Windows passes command line arguments.  On Linux, you get a bunch of strings via the `execve(2)` function that “starts” your program.  And I see no reason to assume otherwise for Windows, at least [the C `main` function][1] uses `int argc, char *argv[]`.  The way you've described the PowerShell matches what `bash` does.

I would rather consider “list of C-strings” via `getArgs` to be the interface from the caller to `stack`, assuming that each OS provides the means to conveniently serve this interface (`bash`, *PowerShell*).

I would object here:

> It is complicated, principally because you have to know what your shell is doing with the command line

Actually, no.  I think `stack` should not try to be smart about what the shell does.  The operating system delivers the list of strings, and that's that.  `stack` should just pass on a section of that list.

The crux with the current approach is, that *individual tokens* of the original list of arguments are split into further lists.  This leads to users fighting this internal logic, having to combine the shell's quoting mechanism with another quoting mechanism, hich does not coincide with the logic of how their shell thinks about lists of arguments.

The data model “one flat list of tokens” does not provide a means for nesting.  But we want nested lists, for `--ta` and its kind.  So there are two ways to work around this:

  1. Encode a list in one token.  Currently done.  Implies additional logic to tokenise the encoded list.  Leads to the trouble we see.

  2. Use tokens to delimit lists.  This is a veteran aproach: It is exactly what we do when parsing a programming language.  The sequence of tokens (characters, XML-tags) is flat, but we give some of them the meaning to start or end nesting, be it `(·)` or `<a>…</a>`.

And this is what I've proposed: 

    $ stack test --ta-x one 'two words' x  --file-watch

The tokens `--ta-x` and `x` take on the roles of parenthesis, the section [`one`, `two words`] is passed on unchanged

Hmm-hm.  I was having a look at optparse-applicative this evening.  I have to admit I'm a bit lost.  I currently don't see how it could be used to implement this.

Sorry.  Consider my suggestion above a wish, but I think I cannot do it with `optparse-applicative`.

(Note to my future self: Remain wary of command line parsing libraries.  They tend to make quite strong assumptions: The datamodel to parse into, their own notions of “flags”, “options” and “arguments”, and their own ideal of command line interface best practice.  There's consensus for none of these assumptions.  If you consider a library to handle CLI, you may be overengineering…)

[1]: https://learn.microsoft.com/en-us/cpp/c-language/parsing-c-command-line-arguments?view=msvc-170
